OPHD/UI/StringTable.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <NAS2D/Renderer/Renderer.h>
+#include <NAS2D/Renderer/Color.h>
+#include <NAS2D/Renderer/Point.h>
+#include <string>
+#include <vector>
+#include <cstddef>
+
+// Draw a 2 dimensional table of text. Determine cell size based on inserted text, font, and padding. Only allows one line of text per cell.
+class StringTable
+{
+public:
+    using CellCoordinate = NAS2D::Point<std::size_t>;
+
+    enum class Justification
+    {
+        Left,
+        Right
+    };
+
+    // Set default fonts in constructor
+    StringTable(std::size_t columns, std::size_t rows);
+
+    void draw(NAS2D::Renderer& renderer) const;
+
+    void setPosition(NAS2D::Point<float> position);
+    NAS2D::Point<float> getPosition() const;
+
+    void setDefaultFont(NAS2D::Font* font);
+    void setDefaultTitleFont(NAS2D::Font* font);
+    void setDefaultTextColor(NAS2D::Color textColor);
+
+    void setHorizontalPadding(float horizontalPadding);
+    void setVerticalPadding(float verticalPadding);
+
+    void setCellText(std::size_t column, std::size_t row, std::string text);
+    void setCellText(const CellCoordinate& cellCoordinate, std::string text);
+    // Override default font settings
+    void setCellFont(const CellCoordinate& cellCoordinate, NAS2D::Font* font);
+    void setCellJustification(const CellCoordinate& cellCoordinate, Justification justification);
+    void setColumnJustification(std::size_t column, Justification justification);
+    // Override default Color setting
+    void setCellTextColor(const CellCoordinate& cellCoordinate, NAS2D::Color textColor);
+
+    // Call after updating table properties to recompute cell positions
+    void computeRelativeCellPositions();
+
+private:
+    struct Cell
+    {
+        // Set textColor to ColorEmpty to indicate cell should use default StringTable color
+        static const NAS2D::Color ColorEmpty;
+
+        // Use defaultFont if not set
+        NAS2D::Font* font = nullptr;
+        std::string text;
+        Justification justification = Justification::Left;
+        NAS2D::Color textColor = ColorEmpty;
+        // Position relative to the StringTable's position
+        NAS2D::Point<float> textRelativePosition;
```

Comment:
A "relative" position should generally be a `Vector` rather than a `Point`.

Reply:
Okay. I could use some help understanding the difference and use cases. I was having trouble with Point when trying to add 2 points together.

Reply:
A `Point` represents a position, while a `Vector` represents a displacement.

Addition is supported in the following way:
- `Vector + Vector ==> Vector`
- `Point + Vector ==> Point` (equivalently `Vector + Point ==> Point`, assuming it's defined)
- `Point + Point` invalid

Similarly, for subtraction:
- `Vector - Vector ==> Vector`
- `Point - Vector ==> Point`
- `Point - Point ==> Vector`
- `Vector - Point` invalid (unless you define `-Point` and treat it as `-Point + Vector`)

Of course it does make geometric sense to average points:
- `(Point + Point) / 2 ==> Point`
- `(Point + Point + Point) / 3  => Point`
- ...

----

One way to represent this is to add a 3rd coordinate, which is 0 for vectors and 1 for points:
- `(x1, y1, 0) + (x2, y2, 0) ==> (x1 + x2, y1 + y2, 0)` (Vector)
- `(x1, y1, 1) + (x2, y2, 0) ==> (x1 + x2, y1 + y2, 1)` (Point)
- `(x1, y1, 1) + (x2, y2, 1) ==> (x1 + x2, y1 + y2, 2)` (invalid)

Of course, since that 3rd coordinate is basically encoded in the type system, and statically checked by the compiler, there is no need to allocate any memory for it, nor spend any CPU cycles calculating it.

Works for point averaging as well:
- `((x1, y1, 1) + (x2, y2, 1)) / 2  ==> ((x1 + x2) / 2, (y1 + y2) / 2, 1) ` (Point)

Reply:
Ah yes, you cannot add two locations together as that wouldn't make any since. Thanks for the math refresher, I'm a bit embarrassed I didn't recognize this up front.

In MonoGame (XNA), this behaviour is allowed:

```CSharp
Point point1 = new Point(0, 1);
Point point2 = new Point(1, 1);
Point point3 = point1 + point2;
```

I guess while logically incorrect to add two points (location), the end result works for displaying graphics on the screen, collision avoidance, etc. I had just never realized I was doing it incorrectly in other projects.

I'll try to fix this up today and use NAS2D::Point for the position and NAS2D::Vector for the relative position (displacement) of each cell.